sql/clients_last_seen_raw_v1.sql

```diff
@@ -0,0 +1,135 @@
+  -- Equivalent to, but more efficient than, calling udf_bitmask_range(1, 28)
+CREATE TEMP FUNCTION
+  udf_bitmask_lowest_28() AS (0x0FFFFFFF);
+  -- Equivalent to, but more efficient than, calling udf_bitmask_range(1, 7)
+CREATE TEMP FUNCTION
+  udf_bitmask_lowest_7() AS (0x7F);
+  --
+CREATE TEMP FUNCTION
+  udf_shift_one_day(x INT64) AS (IFNULL((x << 1) & udf_bitmask_lowest_28(),
+      0));
+  --
+CREATE TEMP FUNCTION
+  udf_combine_days(prev INT64,
+    curr INT64) AS (udf_shift_one_day(prev) + IFNULL(curr,
+      0));
+  --
+CREATE TEMP FUNCTION
+  udf_combine_experiment_days(
+    --
+    prev ARRAY<STRUCT<experiment STRING,
+    branch STRING,
+    bits INT64>>,
+    --
+    curr ARRAY<STRUCT<experiment STRING,
+    branch STRING,
+    bits INT64>>) AS (
+    -- The below is logically a FULL JOIN, but BigQuery returns error
+    -- "Array scan is not allowed with FULL JOIN" so we have to do two
+    -- separate scans.
+    ARRAY_CONCAT(
+      -- Experiments present in prev (and potentially in curr too)
+      ARRAY(
+      SELECT
+        AS STRUCT experiment,
+        branch,
+        udf_combine_days(prev.bits,
+          curr.bits) AS bits
+      FROM
+        UNNEST(prev) AS prev
+      LEFT JOIN
+        UNNEST(curr) AS curr
+      USING
+        (experiment,
+          branch)
+      WHERE
+        udf_combine_days(prev.bits,
+          curr.bits) > 0),
+      -- Experiments present in curr only
+      ARRAY(
+      SELECT
+        AS STRUCT experiment,
+        branch,
+        curr.bits AS bits
+      FROM
+        UNNEST(curr) AS curr
+      LEFT JOIN
+        UNNEST(prev) AS prev
+      USING
+        (experiment,
+          branch)
+      WHERE
+        prev IS NULL)));
+  --
+WITH
+  _current AS (
+  SELECT
+    -- In this raw table, we capture the history of activity over the past
+    -- 28 days for each usage criterion as a single 64-bit integer. The
+    -- rightmost bit represents whether the user was active in the current day.
+    CAST(TRUE AS INT64) AS days_seen_bits,
+    -- For measuring Active MAU, where this is the days since this
+    -- client_id was an Active User as defined by
+    -- https://docs.telemetry.mozilla.org/cookbooks/active_dau.html
+    CAST(scalar_parent_browser_engagement_total_uri_count_sum >= 5 AS INT64) AS days_visited_5_uri_bits,
+    CAST(devtools_toolbox_opened_count_sum > 0 AS INT64) AS days_opened_dev_tools_bits,
+    -- We only trust profile_date if it is within one week of the ping submission,
+    -- so we ignore any value more than seven days old.
+    IFNULL(1 << DATE_DIFF(submission_date_s3, SAFE.PARSE_DATE("%F",
```

Comment:
since we're doing this and the combine in two places, it seems like it would be worth extracting them to UDFs